src/ORM/DB.php

```diff
@@ -680,4 +813,63 @@ public static function alteration_message($message, $type = "")
     {
         DB::get_schema()->alterationMessage($message, $type);
     }
+
+    /**
+     * Get the name of the database connection to use for the given SQL query
+     * The 'default' connection can be either the primary or a replica connection if configured
+     */
+    private static function getDefaultConnectionName(string $sql = ''): string
+    {
+        if (DB::$mustUsePrimary || DB::$withPrimaryCount > 0 || !DB::hasReplicaConfig()) {
+            return DB::CONN_PRIMARY;
+        }
+        if (DB::isMutableSql($sql)) {
+            DB::$mustUsePrimary = true;
+            return DB::CONN_PRIMARY;
+        }
+        if (DB::$replicaConfigKey) {
+            return DB::$replicaConfigKey;
+        }
+        $name = DB::getRandomReplicaConfigKey();
+        DB::$replicaConfigKey = $name;
+        return $name;
+    }
+
+    /**
+     * Check if the given SQL query is mutable
+     */
+    private static function isMutableSql(string $sql): bool
+    {
+        $dbClass = DB::getConfig(DB::CONN_PRIMARY)['type'];
+        // This must use getServiceSpec() and not Injector::get/create() followed by
+        // getConnector() as this can remove the dbConn from a different connection
+        // under edge case conditions
+        $dbSpec = Injector::inst()->getServiceSpec($dbClass);
+        $connectorService = $dbSpec['properties']['connector'];
+        $connector = Injector::inst()->convertServiceProperty($connectorService);
+        return $connector->isQueryMutable($sql);
+    }
+
+    /**
+     * Get a random replica database configuration key from the available replica configurations
+     * The replica choosen will be used for the rest of the request, unless the primary connection
+     * is forced
+     */
+    private static function getRandomReplicaConfigKey(): string
+    {
+        $replicaNumbers = [];
+        for ($i = 1; $i < DB::MAX_REPLICAS; $i++) {
```

Comment:
```suggestion
        for ($i = 1; $i <= DB::MAX_REPLICAS; $i++) {
```
We need to allow `99` as well.

Reply:
Updated